rfcs/text/0011_persistable_state_service.md

```diff
@@ -0,0 +1,340 @@
+
+# What saved object migrations 
+
+With saved object migrations (or any other migration of potentially high amount of state objects) currently we try to detect when a migration needs to be performed. For example if there was no change in specific saved object between minors we will not try to read or write those objects from/to elasticsearch.
+
+With extention points to which 3rd party developers might register their own items which expose state that is not possible. For example dashboard saved object might contain state from any embeddable. As we don't know about all those embeddables we need to always fetch all saved objects and run migrations on all of them. 
+If expecting to handle large amount of state objects you should always deep compare input and output state and filter out objects without changes to avoid to many updates in the database.
+
+
+# Full look into a complex example: saved dashboards
+
+saved dashboard state looks something like this:
+```ts
+{
+  panels: [{
+   ...embeddable,
+   enhancements: {
+    drilldowns: {
+      ....drilldownStateForThisEmbeddable
+    }
+  } 
+}],
+}
+```
+
+where any plugin can add to enhancements a property name matching its state id with value being that state.
+also any embeddable can extend the base embeddable input type.
+
+dashboard needs to register a persistable state service with migrate function looking something like this:
+```ts
+const migrate = (state: unknown, version: string) => {
+  // do the internal state migration like state.title etc (owned by dashboard)
+  
+  // and then migrate panels (state not owned by dashboard)
+  state.panels = state.panels.map(p => baseEmbeddablePesistableStateDefinition.migrate(p, version));
+}
+```
+
+all panels extend base embeddable, which would register its baseEmbeddablePErsistableState with migrate function like this:
+
+```ts
+const migrate = (state: unknown, version: string) => {
+  // do the internal state migration like state.title etc (owned by base embeddable)
+  
+  // then migrate enhancements where every key represents a state not owned by embeddable
+  Object.keys(state.enhancements).forEach(k => {
+    state.enhancements[k] = persistableStateRegistry.get(k).migrate(state.enhancements[k], version);
+  }
+  
+  // migrate the actual embeddable type
+  // here we run into an issue as base embeddable input and specific embeddable input should not collide (thru any version) ... 
+  // this is mentioned before "We have to avoid shared persistable state" topic
+  // we should make all embeddables provide their custom inputs under a base property to prevent that
+  return persistableStateRegistry.get(state.type).migrate(state, version),  
```

Comment:
Going back on the 

> A plugin could have a single `stateType`, for example `visualizations` plugin has a `savedVis` state, so it use `visualizations-savedVis` as id. Other plugins might have multiple state types they want to register. Lets say that we introduce a new state type to `visualizations` plugin called `visState` we would use `visualizations-visState` as id.

part here. I guess that's not the case atm (types are not prefixed with the pluginId, are they). How will 'premigration' of the existing persisted data be performed?

Reply:
very good point, i think we need to allow migrate function to return different state id. so when i would be migrating the current dashboard enhancements (`enhancement.drilldown`) i can return new state id and change that to `enhancements.drilldowns_drilldownState`